### PR TITLE
Add Darwin/macOS support for SSS and Tang pins

### PR DESCRIPTION
**Pull request by sini**
## Summary

This adds macOS/Darwin compatibility to the SSS and Tang pins by replacing
Linux-specific APIs with POSIX equivalents and making the LUKS test suite
gracefully skip when cryptsetup is unavailable.

Tested on macOS 26 (aarch64-darwin) — Tang and SSS-based encryption and
decryption both work correctly.

This may also help with support for #541 and #504 

## Changes

**sss: remove unused `sys/epoll.h` include from `clevis-encrypt-sss.c`**
This header was included but never used. Removing it fixes a compile error on
Darwin where `sys/epoll.h` does not exist.

**sss: replace Linux `epoll` with POSIX `poll` in `clevis-decrypt-sss.c`**
`epoll` is Linux-specific. The SSS decrypt pin monitors a small number of child
process file descriptors, so `poll()` is functionally equivalent and is available
on all POSIX platforms.

**sss: replace `pipe2` with portable `pipe` + `fcntl` in `sss.c`**
`pipe2(fd, O_CLOEXEC)` is Linux-specific. The replacement uses `pipe()` followed
by `fcntl(F_SETFD, FD_CLOEXEC)` on both descriptors. This is safe because the
`call()` function operates in a single-threaded context (the fork follows
immediately), so there is no window for a descriptor leak.

**luks: make `cryptsetup` optional in test suite**
The LUKS test `meson.build` is included unconditionally by the parent build, but
all tests require `cryptsetup`, which is unavailable on macOS. Changed
`find_program('cryptsetup', required: true)` to `required: false` with an early
`subdir_done()` so the build completes without cryptsetup. No test logic is
altered — when cryptsetup is present, all tests run as before.

## Motivation

Clevis is used in NixOS disk encryption workflows (Tang + SSS) to generate and
decrypt JWE-wrapped keys. Being able to run `clevis encrypt tang` and
`clevis encrypt sss` on macOS enables Darwin-based workstations to provision
NixOS hosts with encrypted disks without requiring a Linux VM.

## Test plan

- [x] Builds on x86_64-linux (NixOS)
- [x] Builds on aarch64-darwin (macOS 26)
- [x] `clevis encrypt tang` / `clevis decrypt` round-trip on macOS
- [x] `clevis encrypt sss` / `clevis decrypt` round-trip on macOS
- [x] Existing Linux build and test suite unaffected (no behavior change when
      cryptsetup is present)

### CODE REVIEW

**Comment by sini** [comment]
I'll take a look and see if I can reproduce the timeout test failure locally.

EDIT: I reproduced it and think I have the root cause -- working on the fix, moving this to draft until I've resolved it.

**Comment by sini** [comment]
The epoll to poll conversion had a subtle bug with fd lifecycle management. When epoll monitors a closed fd, it automatically removes it from the set. Poll doesn't - it keeps returning POLLHUP indefinitely, causing an infinite loop. 

Why it wasn't caught initially: I only tested on Darwin (which was the whole point of the port), where the Tang/SSS functionality worked fine for my use case. That was an oversight on my part, I've now verified the whole suite works on all tested platforms. The bug only surfaced when running the full Linux test suite, which exercised edge cases like child process failures and invalid configurations. 

The fix:
- Set closed fds to -1 so poll ignores them (matching epoll's auto-removal)
- Handle POLLHUP/POLLERR/POLLNVAL to clean up failed children - but only when there's no POLLIN, since a process can write valid data and immediately exit (setting both POLLHUP and POLLIN). Reading the data first ensures we don't discard valid output.

**Comment by sergio-correia** [comment]
@sini: Hello, could you please rebase this on top of the current tree? Thanks in advance.

**Comment by sini** [comment]
@sergio-correia Sorry, I missed this notification. I'll rebase and update it now!

**Review by sarroutbi — CHANGES_REQUESTED** [review]
## Synchronous `waitpid()` introduces a potential deadlock/hang

### Location

[`src/pins/sss/clevis-decrypt-sss.c`](src/pins/sss/clevis-decrypt-sss.c) — introduced in commit `74b7961` (*sss: handle poll events correctly in clevis-decrypt-sss*)

### Description

The error handler for `POLLERR | POLLHUP | POLLNVAL` calls `waitpid(pin->pid, NULL, 0)` to reap the child process:

```c
/* If no data available but pipe closed/errored, mark as failed */
if (!(pollfds[pi].revents & (POLLIN | POLLPRI))) {
    if (pollfds[pi].revents & (POLLERR | POLLHUP | POLLNVAL)) {
        fclose(pin->file);
        pin->file = NULL;
        pollfds[pi].fd = -1;
        waitpid(pin->pid, NULL, 0);   // <-- blocking call
        pin->pid = 0;
        pin->next->prev = pin->prev;
        pin->prev->next = pin->next;
        free(pin);
        break;
    }
    continue;
}
```

This is a **blocking call** (flags argument is `0`). `poll()` flags `POLLHUP` as soon as the child closes its end of the pipe (stdout). If a child closes its output but continues executing (or gets stuck), `waitpid` blocks the main thread indefinitely. This stops the entire event loop, starving all other healthy child processes and hanging the decryption process.

### Affected paths

The same issue exists on the **success path** as well (line 280):

```c
fclose(pin->file);
pin->file = NULL;
pollfds[pi].fd = -1;

waitpid(pin->pid, NULL, 0);   // <-- also blocking
pin->pid = 0;
```

Both call sites block synchronously inside the event loop.

### Practical impact

For typical clevis pin processes (short-lived decrypt commands), the window between closing stdout and exiting is negligible, so the risk is low in practice. However, this is a correctness bug in general — any misbehaving or slow-to-exit child will hang the entire SSS decryption.

### Recommendation

- Use `WNOHANG` and let the cleanup loop handle final reaping:

Replace both `waitpid` calls inside the event loop with:

```c
waitpid(pin->pid, NULL, WNOHANG);
```

The `egress` cleanup (lines 337-352) already sends `SIGTERM` and performs a blocking `waitpid` on any remaining children, so zombies will not leak.

**Review by sini — COMMENTED** [review]
Thanks for the thorough review @sarroutbi! I've addressed the `waitpid` concern, but went a different direction than `WNOHANG` — here's my reasoning:

**`waitpid` — kept blocking (`0`) instead of `WNOHANG`:**

Using `WNOHANG` introduces a zombie risk: if the child hasn't called `_exit()` yet when we call `waitpid`, it returns 0 without reaping. We then set `pin->pid = 0` and free the pin, so the `egress` cleanup (which only kills/waits on `pid > 0`) can never reap it — the child becomes a permanent zombie.

At both call sites the pipe has already been closed or fully drained (`fclose` + `fread` until EOF on the success path, `fclose` on the POLLHUP path), so a blocking wait is safe — there's no deadlock risk since there's no shared pipe that could cause the child to block on write. The child is either already exited or in its final teardown, so the blocking wait returns nearly instantly in practice.

If a misbehaving child hangs after closing stdout, that's arguably the correct behavior — we should wait for it rather than silently orphan it. The `egress` path already handles the abort case with `SIGTERM` + blocking wait.

** Improvements in this update: **

1. **`fcntl` error checking** — the `pipe()` + `fcntl(F_SETFD, FD_CLOEXEC)` replacement for `pipe2` now checks `fcntl` return values and jumps to the error cleanup on failure.

**Comment by sini** [comment]
**Update:** Removed the `EINTR` handling from my previous push — that was scope creep. The original `epoll_wait` loop didn't handle `EINTR` either (`if (r != 1) break`), and this is a short-lived decrypt process, not a long-running daemon. The poll loop now matches the original semantics: `if (r <= 0) break`.

The `fcntl` error checking fix in `sss.c` is still included — that one is directly related to the `pipe2` → `pipe` + `fcntl` conversion.

Current state of the branch:
1. `34e8a5c` — remove unused `sys/epoll.h` from encrypt
2. `9130e28` — replace epoll with poll in decrypt
3. `7b84f01` — replace `pipe2` with `pipe` + `fcntl` (with error checking)
4. `58da0cd` — make cryptsetup optional in test suite
5. `4f0a02e` — handle POLLHUP/POLLERR/POLLNVAL correctly in poll loop

**Comment by sini** [comment]
@sarroutbi what would you like to do here?

**Comment by sarroutbi** [comment]
> @sarroutbi what would you like to do here?

Hi @sini,

Great catch on the zombie risk with `WNOHANG`—clearing the PID before the process fully exits would definitely leak resources. Also, verifying the `fcntl` return values for the Darwin fallback is a great improvement for stability.

That being said, keeping `waitpid(..., 0)` as a synchronous blocking call inside the main event loop is dangerous. Even if the pipe is drained, a child could still hang during its internal destruction (e.g., stuck in a signal handler, library destructor, or an internal deadlock). If that happens, the blocking wait will freeze the entire Clevis process, breaking the concurrency of the `poll` architecture.

If we want to keep the loop 100% non-blocking without micro-managing zombie states, we could leverage the OS kernel before forking by setting `SIGCHLD` with `SA_NOCLDWAIT`. This tells the kernel to automatically reap dead children without requiring a synchronous `waitpid` from the parent. 

If that feels out of scope for this Darwin port, a simple deferred cleanup list for active PIDs might be a safer middle ground than a blocking wait.

Let me know what you think!

**Comment by sini** [comment]
@sarroutbi You're right that a blocking `waitpid` has no business in the middle of the poll loop — a child that stalls after closing stdout shouldn't be able to freeze the whole process and starve its siblings. I went with your deferred-cleanup-list suggestion rather than `SA_NOCLDWAIT`, for two reasons:

- `SA_NOCLDWAIT` changes reaping globally but doesn't actually bound the case you described: a child stuck in a destructor/handler is still *alive*, so nothing reaps it and nothing is waiting — we'd just lose the ability to `SIGTERM` it on abort. And its interaction with a later `wait()` is left partly unspecified by POSIX (a `waitpid` with live children may block until *all* exit, then fail `ECHILD`), which is exactly the kind of cross-platform surprise this port is trying to avoid.
- A deferred PID list keeps the reaping explicit and portable, and lets us bound the pathological case with `SIGTERM` at teardown.

What changed (`clevis-decrypt-sss.c`):

- Both in-loop `waitpid(..., 0)` sites (the `POLLHUP`/error path and the success path) now push the pid onto a `reap[]` list via `defer_reap()` instead of blocking. **The poll loop is now 100% non-blocking.**
- `egress` drains that list with `kill(SIGTERM)` + `waitpid`, mirroring the existing aborted-child cleanup. Blocking there is safe — the loop has already finished, so a slow child can't affect its siblings, and the `SIGTERM` bounds anything that stalls after closing stdout.
- No zombies (the `WNOHANG` failure mode), no reliance on signal-disposition inheritance across `fork`, and the `nchldrn` accounting / all-negative-fd poll guard are untouched.

Full `sss` suite passes (success, single-child-fail `POLLHUP`, all-fail give-up, and the `t=2` multi-child variants all exercise the changed paths). Pushed as a separate commit on top.

File: src/pins/sss/clevis-decrypt-sss.c [hunk 1]
```diff
@@ -195,31 +212,59 @@ main(int argc, char *argv[])
...
         goto egress;
 
-    for (struct epoll_event e; true; ) {
-        int r = 0;
-
-        r = epoll_wait(epoll, &e, 1, -1);
-        if (r != 1)
+    while (true) {
+        int r = poll(pollfds, nfds, -1);
+        if (r <= 0)
```
Comment (sergio-correia) [inline src/pins/sss/clevis-decrypt-sss.c]: `poll()` returns -1 with `errno == EINTR` when interrupted by a signal, but the `r <= 0` check treats it as terminal and breaks the loop. In initramfs/dracut boot environments, inherited signal handlers (watchdog `SIGALRM`, `SIGWINCH`, etc.) can trigger this, silently aborting SSS decryption even though all children are healthy.

This is pre-existing (the original `epoll_wait` had the same semantics), but since the poll loop is being rewritten anyway, would it be worth adding the standard EINTR retry?

```suggestion
        int r;
        do {
            r = poll(pollfds, nfds, -1);
        } while (r < 0 && errno == EINTR);
        if (r <= 0)
```

File: src/pins/sss/clevis-decrypt-sss.c [hunk 2]
```diff
@@ -131,6 +131,24 @@ compact_jwe(FILE *file)
...
+ * poll loop non-blocking: a child that stalls after closing its stdout can
+ * no longer freeze the event loop and starve its healthy siblings. */
+static void
+defer_reap(pid_t **reap, size_t *nreap, pid_t pid)
+{
+    pid_t *tmp = realloc(*reap, (*nreap + 1) * sizeof(*tmp));
+
+    if (!tmp) {
+        /* Out of memory: reap now rather than leak the child. */
+        waitpid(pid, NULL, 0);
```
Comment (sergio-correia) [inline src/pins/sss/clevis-decrypt-sss.c]: The whole point of `defer_reap()` is to keep the poll loop non-blocking, but the OOM fallback calls `waitpid(pid, NULL, 0)` which re-introduces the exact blocking pattern this function was designed to eliminate. If `realloc` fails while a child is stuck in a destructor after closing stdout, the loop freezes.

Granted, OOM on a few-element `pid_t` array is extremely unlikely, but using `WNOHANG` here would keep the invariant consistent. A zombie leak under OOM is acceptable since `clevis-decrypt-sss` is short-lived (init adopts the child at exit):

```suggestion
        waitpid(pid, NULL, WNOHANG);
```

File: src/pins/sss/clevis-decrypt-sss.c (hunk 1 above)
Comment (sini) [reply src/pins/sss/clevis-decrypt-sss.c]: Added, in `2a7a5f2`.

One correction to the stated mechanism, since it changes how reachable this is today: handlers inherited into initramfs can't be the trigger. `execve()` resets every *caught* signal to `SIG_DFL` (ignored dispositions survive, but `SIG_IGN` never produces `EINTR`), so a watchdog `SIGALRM` handler installed by dracut is gone by the time `clevis-decrypt-sss` reaches `poll()`. `EINTR` requires a handler in *this* process, and there isn't one.

That said, the retry belongs there. "No handler exists" is a property of today's code, not an invariant anyone is maintaining — the first `signal()`/`sigaction()` call added to this binary would silently turn every delivery into an aborted decryption, and the failure would be intermittent and boot-time. Cheap to make unconditional:

```c
        do {
            r = poll(pollfds, nfds, -1);
        } while (r < 0 && errno == EINTR);
        if (r <= 0)
            break;
```

With an infinite timeout `poll()` never returns 0, so `r <= 0` after the retry is now exactly the genuine-error case.

File: src/pins/sss/clevis-decrypt-sss.c (hunk 2 above)
Comment (sini) [reply src/pins/sss/clevis-decrypt-sss.c]: You're right that this branch falsifies the invariant — `defer_reap()` exists to keep the loop non-blocking and then blocks in its own fallback. Fixed in `2a7a5f2`, though not with `WNOHANG`.

`WNOHANG` keeps the failure mode alive, just swapping which end it hurts: if the child hasn't exited yet we return without reaping, the pid is dropped from the list, and nothing can ever reap it. Both outcomes are a choice between a hang and a leak, which is only necessary because the allocation can fail at all.

It doesn't have to. The child count is known before the first `fork()`, and each child is queued exactly once, so the list can be sized up front and never grows inside the loop:

```c
    npins = json_array_size(pins);
    if (npins > 0) {
        reap = calloc(npins, sizeof(*reap));
        if (!reap)
            goto egress;
    }
```

`defer_reap()` reduces to `reap[(*nreap)++] = pid;` — no allocation, no failure path, nothing to decide under OOM. Allocation failure now happens before any child exists, where `egress` already does the right thing.

While checking this I hit a separate hang in the same loop, pushed as `1effa22`: a JWE with an empty `clevis.sss.jwe` array spawns no children, so `poll(NULL, 0, -1)` blocks forever, and the `nchldrn() < t` check that would catch it sits after the poll. It predates this PR — `epoll_wait()` on an empty set blocks identically — but it's a one-line guard in code this PR already touches. Happy to split it into its own PR if you'd rather keep this one to the Darwin port.

Build is clean under `-Wall -Wextra -Werror`; `pin-sss` and `pin-null` both pass (the `t=1`/`t=2` success, single-child-`POLLHUP`, and all-fail paths all exercise the changed code).

File: src/pins/sss/clevis-decrypt-sss.c [hunk 3]
```diff
@@ -249,8 +308,10 @@ main(int argc, char *argv[])
...
             fclose(pin->file);
             pin->file = NULL;
+            /* Remove closed fd from poll set (poll ignores negative fds) */
+            pollfds[pi].fd = -1;
 
-            waitpid(pin->pid, NULL, 0);
+            defer_reap(reap, &nreap, pin->pid);
             pin->pid = 0;
 
             if (!pin->pt) {
```
Comment (sergio-correia) [inline src/pins/sss/clevis-decrypt-sss.c]: This `if (!pin->pt)` block is now unreachable. In the old epoll code it handled the case where `epoll_wait` delivered `EPOLLHUP` without `EPOLLIN` (the read block was skipped, so `pin->pt` stayed NULL). The new code handles that case separately in the `POLLERR`/`POLLHUP` handler at lines 266-278, so the only way to reach line 317 is through a successful data read where `pin->pt` is non-NULL. All failure paths exit via `goto egress`.

Would you mind either removing this dead block or adding a short comment noting it is defensive dead code? As-is it could mislead someone into thinking the data-read block can fail without hitting `goto egress`.

File: src/luks/tests/meson.build [hunk 4]
```diff
@@ -1,9 +1,14 @@
 # We use jq for comparing the pin config in the clevis luks list tests.
 jq = find_program('jq', required: false)
 
-# We use cryptsetup for testing LUKS2 binding and saving the token in a
-# given token slot.
-cryptsetup = find_program('cryptsetup', required: true)
+# All LUKS tests require cryptsetup. The test directory is included
+# unconditionally by the parent meson.build, so we must handle the case
+# where cryptsetup is not available (e.g., on macOS/Darwin).
+cryptsetup = find_program('cryptsetup', required: false)
```
Comment (sergio-correia) [inline src/luks/tests/meson.build]: This change from `required: true` to `required: false` affects all platforms, not just macOS. On Linux, if a CI container omits `cryptsetup`, meson will now silently skip all LUKS tests with a `warning()` instead of failing at configure time. The original `required: true` was a deliberate assertion.

Would it be better to gate this on the host OS so Linux builds keep the hard requirement?

```suggestion
cryptsetup = find_program('cryptsetup', required: host_machine.system() == 'linux')
```

**Review by sergio-correia — COMMENTED** [review]
A few suggestions below. The epoll-to-poll conversion, deferred reaping, and EINTR retry all look correct. Nice catch on the empty descriptor set hang and the pre-allocated reap array.

Not in this diff, but related: `_GNU_SOURCE` is defined in all three SSS files (`clevis-decrypt-sss.c:38`, `clevis-encrypt-sss.c:38`, `sss.c:38`) but is no longer needed now that `pipe2()` and `epoll_create1()` are gone. Removing it would align with the portability goal of this PR.

File: src/pins/sss/clevis-decrypt-sss.c (hunk 3 above)
Comment (sini) [reply src/pins/sss/clevis-decrypt-sss.c]: Confirmed, and removed in `9061c8f` rather than commented.

Agreed on the reachability argument. Every path into line 317 now goes through the successful-read block: the `malloc` is checked with `goto egress`, and the short-read case (`pin->ptl != ptl`) frees `pin->pt`, sets it to NULL and also exits via `goto egress`. Nothing falls through with `pin->pt == NULL`, so the unlink was unreachable.

Dropped it outright — a defensive block that can't fire still has to be reasoned about every time someone reads the loop, and the invariant it was protecting (a pin left in the list with no plaintext) is now enforced structurally by the `POLLERR`/`POLLHUP`/`POLLNVAL` branch owning that case.

Rebuilt under the project flags (`-Wall -Wextra -Werror`, `-std=c99`) and `pin-sss`/`pin-null` both pass, including the `t=1` partial-failure and all-fail paths that used to be the callers of interest for this branch.

File: src/luks/tests/meson.build (hunk 4 above)
Comment (sini) [reply src/luks/tests/meson.build]: Applied in `8883f7b`.

You're right that `required: false` silently widened the change past Darwin. The original `required: true` was an assertion that a Linux build without `cryptsetup` is a broken environment, not a reduced test matrix, and downgrading it to a `warning()` would have turned a configure-time failure into a silently green CI run with every LUKS test skipped.

`host_machine.system() == 'linux'` keeps that assertion exactly where it was and confines the relaxation to the platform that needs it.

**Comment by sini** [comment]
@sergio-correia All three points addressed; pushed as `9061c8f` (dead block), `d01e7a1` (`_GNU_SOURCE`), with the `meson.build` suggestion already in as `8883f7b`. Replies in the two threads.

On `_GNU_SOURCE` — agreed, and removed from all three SSS files.

Checked that nothing else in them needs it: `pipe2()` and `epoll_create1()` were the only GNU-only interfaces, and no `asprintf`/`strdup`/`memmem`/`getline`/`strchrnul`/`mkostemp`/`ppoll` call took their place. Everything remaining is covered by the project-wide `-D_DEFAULT_SOURCE -D_POSIX_C_SOURCE=200112L`. Verified rather than argued: all three compile clean on glibc under the full project flag set, and `-Wimplicit-function-declaration` is fatal there via `-Werror`, so a hidden POSIX declaration would have failed the build rather than degraded silently.

Worth naming the portability argument explicitly, since it's stronger than "unused": `_GNU_SOURCE` is a glibc feature-test macro that Apple's libc ignores outright, so its presence never affected the Darwin build one way or the other. What it did do was keep the door open — a GNU-only call added to these files would compile clean on Linux and only break at the macOS build, which is exactly the failure this PR exists to close. Removing it makes Linux the place that catches it.

One adjacent leftover I noticed while checking, not touched: `clevis-decrypt-sss.c:47` and `clevis-encrypt-sss.c:47` still `#include <libgen.h>`, but neither file calls `basename()` or `dirname()` any more. Happy to drop both lines if you want it in this PR, or leave it out to keep the diff scoped.

`pin-sss` and `pin-null` pass against the rebuilt binaries.
